Moving getting the current state above the GCE specific part does introduce one complication which the tests caught.
```
=== RUN   TestGCECondition
--- FAIL: TestGCECondition (0.00s)
    node_controller_test.go:374: 
                Error Trace:    node_controller_test.go:374
                Error:          Not equal: 
                                expected: 2
                                actual  : 1
                Test:           TestGCECondition
                Messages:       No new conditions were added for GCE
    node_controller_test.go:383: 
                Error Trace:    node_controller_test.go:383
                Error:          Should be true
                Test:           TestGCECondition
                Messages:       Network Route Condition for GCE not added by external cloud initializer
```
What happens is that the `nodeutil.SetNodeCondition` modifies the node state, so when the request for the current node state is done first it is no longer the current state after setting the node condition. Then when the node is updated later, it removes the conditions that were added in the GCE part.

Either we should leave the GCE code at the top of RetryOnConflict, or the current state has to be refreshed after `SetNodeCondition`, like so:
```go
    err := clientretry.RetryOnConflict(UpdateNodeSpecBackoff, func() error {
            curNode, err := cnc.kubeClient.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
            if err != nil {
                    return err 
            }   
 
            cloudTaint := getCloudTaint(curNode.Spec.Taints)
            if cloudTaint == nil {
                    // Node object received from event had the cloud taint but was outdated,
                    // the node has actually already been initialized.
                    return nil 
            }   
 
            // TODO(wlan0): Move this logic to the route controller using the node taint instead of condition
            // Since there are node taints, do we still need this?
            // This condition marks the node as unusable until routes are initialized in the cloud provider
            if cnc.cloud.ProviderName() == "gce" {
                    if err := nodeutil.SetNodeCondition(cnc.kubeClient, types.NodeName(node.Name), v1.NodeCondition{
                            Type:               v1.NodeNetworkUnavailable,
                            Status:             v1.ConditionTrue,
                            Reason:             "NoRouteCreated",
                            Message:            "Node created without a route",
                            LastTransitionTime: metav1.Now(),
                    }); err != nil {
                            return err 
                    }   
+
+                   var err error
+                   curNode, err = cnc.kubeClient.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
+                   if err != nil {
+                           return err 
+                   }   
            }
```

--- a/pkg/controller/cloud/node_controller.go
+++ b/pkg/controller/cloud/node_controller.go
@@ -256,6 +256,13 @@ func (cnc *CloudNodeController) initializeNode(node *v1.Node) {
 			return err
 		}
 
+		cloudTaint := getCloudTaint(curNode.Spec.Taints)
+		if cloudTaint == nil {
+			// Node object received from event had the cloud taint but was outdated,
+			// the node has actually already been initialized.
+			return nil
+		}
+
 		if curNode.Spec.ProviderID == "" {
 			providerID, err := cloudprovider.GetInstanceProviderID(context.TODO(), cnc.cloud, types.NodeName(curNode.Name))
 			if err == nil {
@@ -312,14 +319,14 @@ func (cnc *CloudNodeController) initializeNode(node *v1.Node) {
 		// After adding, call UpdateNodeAddress to set the CloudProvider provided IPAddresses
 		// So that users do not see any significant delay in IP addresses being filled into the node
 		cnc.updateNodeAddress(curNode, instances)
+
+		klog.Infof("Successfully initialized node %s with cloud provider", node.Name)
 		return nil
 	})
 	if err != nil {
 		utilruntime.HandleError(err)
 		return
 	}
-
-	klog.Infof("Successfully initialized node %s with cloud provider", node.Name)
 }
 
 func getCloudTaint(taints []v1.Taint) *v1.Taint {

